### PR TITLE
Introduce coroutine to GeoJsonSource.kt

### PR DESCRIPTION
**Pull request by Chaoba**
PRs must be submitted under the terms of our Contributor License Agreement [CLA](https://github.com/mapbox/mapbox-maps-android/blob/main/CONTRIBUTING.md#contributor-license-agreement).
Fixes: < Link to related issues that will be fixed by this pull request, if they exist >

## Pull request checklist:
 - [x] Briefly describe the changes in this PR.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
 - [ ] Write tests for all new functionality. If tests were not written, please explain why.
 - [ ] Add example if relevant.
 - [ ] Document any changes to public APIs.
 - [x] Apply changelog label ('breaking change', 'bug :beetle:', 'build', 'docs', 'feature :green_apple:', 'performance :zap:', 'testing :100:') or use the label 'skip changelog'
 - [x] Add an entry inside this element for inclusion in the `mapbox-maps-android` changelog: `<changelog>Introduce coroutine to GeoJsonSource to remove static worker parsing thread that can't be precisely controlled.</changelog>`.

### Summary of changes
Use coroutine to parse GeoJson data instead of HandlerThread

### User impact (optional)
Following code could be removed from tests:
```
    GeoJsonSource.workerThread =
      HandlerThread("STYLE_WORKER").apply {
        priority = Thread.MAX_PRIORITY
        start()
      }
```

### CODE REVIEW

File: extension-style/build.gradle.kts [hunk 1]
```diff
@@ -23,6 +23,7 @@ dependencies {
   implementation(Dependencies.androidxAppCompat)
   implementation(Dependencies.androidxCoreKtx)
   implementation(Dependencies.androidxAnnotations)
+  implementation(Dependencies.coroutines)
```
Comment (kiryldz) [inline extension-style/build.gradle.kts:26]: Just curious - how does this affect binary size?

File: extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt [hunk 2]
```diff
@@ -594,27 +597,47 @@ class GeoJsonSource(builder: Builder) : Source(builder.sourceId) {
...
-      // remove any events from queue before posting this task
-      workerHandler.removeCallbacksAndMessages(null)
-      workerHandler.post {
-        val property = data.toPropertyValue()
-        mainHandler.post {
-          // we set parsed data when sync setter was not called during background work
-          if (ignoreParsedGeoJsonRegistry[sourceId] == false) {
-            setProperty(property, throwRuntimeException = false)
-            listener.invoke(this)
+      // No parseJob is working now, parse the data directly.
```
Comment (kiryldz) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:600]: In other places `workerHandler.removeCallbacksAndMessages(null)` was replace by `waitingData = null` but in this case some another logic appears.

Comment (Chaoba) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:600]: HandlerThread has a queue and could pop the events automatically, we don't have the queue and need to handle the logic mannually.

File: extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt [hunk 3]
```diff
@@ -41,22 +39,27 @@ class GeoJsonSource(builder: Builder) : Source(builder.sourceId) {
...
           }
         }
       }
-    } ?: run { geoJsonParsed = true }
+    } ?: run {
+      geoJsonParsed = true
+    }
   }
 
+  private suspend fun getPropertyValue(it: GeoJson) =
```
Comment (kiryldz) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:58]: can we replace it with extension function for `GeoJson` and have another function name similar to `parseData`.

File: extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt (hunk 2 above)
Comment (kiryldz) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:600]: my high level q is more about that this manual logic is used only here while in code above is technically almost the same and we do not perform similar logic there.
but I did not give it a deep look - I'll check out your branch later to validate it myself but please focus on retaining logic 1 to 1 as it was before, we need to be careful not to bring in unexpected bugs 😉

File: extension-style/build.gradle.kts (hunk 1 above)
Comment (Chaoba) [reply extension-style/build.gradle.kts:26]: The release aar size for extension-style on master is 612kb, on this branch is 613 kb.

File: extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt (hunk 3 above)
Comment (Chaoba) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:58]: ` GeoJson.toPropertyValue()` is an extension function, this function is just a wrap the function call int default dispatcher. IMO, we don't need another extension function have the same functionality with ` GeoJson.toPropertyValue()`

File: extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt (hunk 2 above)
Comment (Chaoba) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:600]: `code above` here means the part in the constructor?  One instance can't be constructed twice, so the logic for that part can be simpler than this part as this function can be invoked multiple times

File: extension-style/build.gradle.kts (hunk 1 above)
Comment (pengdev) [reply extension-style/build.gradle.kts:26]: I guess the aar doesn't contain the binary size of coroutine dependency 🙂

File: extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt [hunk 4]
```diff
@@ -28,11 +25,12 @@ import com.mapbox.maps.extension.style.utils.toValue
...
  */
 class GeoJsonSource(builder: Builder) : Source(builder.sourceId) {
+  private var parsingJob: Job? = null
   private var geoJsonParsed = false
   private val onGeoJsonParsedListenerList = mutableListOf<OnGeoJsonParsed>()
-  private val workerHandler by lazy {
-    Handler(workerThread.looper)
-  }
+  private val scope = MainScope()
+  private var waitingData: Pair<GeoJson, ((GeoJsonSource) -> Unit)>? = null
```
Comment (pengdev) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:32]: nit: rename to `pendingData`?

**Comment by pengdev** [comment]
The only concern for me is the binary size increase by introducing the `coroutine` dependency, otherwise, as long as the tests pass and examples work as expected. It's a 🟢 for me 🙂

File: sdk/src/androidTest/java/com/mapbox/maps/MapIntegrationTest.kt [hunk 5]
```diff
@@ -33,6 +34,24 @@ class MapIntegrationTest {
...
+    return Point.fromLngLat(
+      random.nextDouble() * -360.0 + 180.0,
+      random.nextDouble() * -180.0 + 90.0
+    )
+  }
+
+  private fun getFeatureCollection(): FeatureCollection {
+    val pointList = mutableListOf<Feature>()
+    for (i in 0..pointCount) {
+      val point = createRandomPoint()
```
[the remark this answers is not in the document]
Comment (Chaoba) [reply sdk/src/androidTest/java/com/mapbox/maps/MapIntegrationTest.kt:50]: That's a good point. But in this case, we only want to verify loading large data, the value in these points actually makes no difference.

File: sdk/src/androidTest/java/com/mapbox/maps/MapIntegrationTest.kt [hunk 6]
```diff
@@ -480,6 +490,105 @@ class MapIntegrationTest {
...
+          .build()
+      )
+      mapboxMap.loadStyleUri(Style.MAPBOX_STREETS) { style ->
+        val source = geoJsonSource("source") {
+          data("")
+        }
+        // load a large feature collection
+        source.featureCollection(getFeatureCollection())
+        // change geometry immediately
+        source.geometry(Point.fromLngLat(0.0, 0.0)) {}
```
Comment (kiryldz) [inline sdk/src/androidTest/java/com/mapbox/maps/MapIntegrationTest.kt:561]: Comment seems to be confusing as we're applying not immediately here but in async manner after large `featureCollection` applied in a sync manner.

File: sdk/src/androidTest/java/com/mapbox/maps/MapIntegrationTest.kt [hunk 7]
```diff
@@ -480,6 +490,105 @@ class MapIntegrationTest {
...
+
+        style.addSource(source)
+        style.addLayer(
+          circleLayer("layer", "source") {
+            circleColor("red")
+            circleRadius(10.0)
+          }
+        )
+        val min = ScreenCoordinate(mapView.width / 2.0 - 10, mapView.height / 2.0 - 10)
+        val max = ScreenCoordinate(mapView.width / 2.0 + 10, mapView.height / 2.0 + 10)
```
Comment (kiryldz) [inline sdk/src/androidTest/java/com/mapbox/maps/MapIntegrationTest.kt:571]: Why do we need a screen box here?
If I understand test correctly we should have exactly one point (0.0, 0.0) as a result

File: extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt [hunk 8]
```diff
@@ -41,22 +39,33 @@ class GeoJsonSource(builder: Builder) : Source(builder.sourceId) {
...
+          onGeoJsonParsedListenerList.forEach {
+            it.onGeoJsonParsed(this@GeoJsonSource)
           }
         }
+        pendingData?.let {
+          // have data waiting to parse, continue parse this data
+          parseData(it.first, it.second)
+          // clear waitingData, so that if new data coming could wait.
+          pendingData = null
+        }
```
Comment (kiryldz) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:57]: This code is duplicated below (except notifying listeners). However it seems very important to me. Can we introduce new private function and re-use it?

File: sdk/src/androidTest/java/com/mapbox/maps/MapIntegrationTest.kt (hunk 7 above)
Comment (Chaoba) [reply sdk/src/androidTest/java/com/mapbox/maps/MapIntegrationTest.kt:571]: This case failed on firebase in the past, but should be ok now, will change it back.

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/CircleAnnotationActivity.kt [hunk 9]
```diff
@@ -69,7 +69,7 @@ class CircleAnnotationActivity : AppCompatActivity() {
 
         // random add circles across the globe
         val circleAnnotationOptionsList: MutableList<CircleAnnotationOptions> = ArrayList()
-        for (i in 0..2000) {
+        for (i in 0..20) {
```
Comment (kiryldz) [inline app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/CircleAnnotationActivity.kt:72]: curious why did we decrease number here?

Comment (Chaoba) [reply app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/CircleAnnotationActivity.kt:72]: Other examples are all 20, 2000 is introduced to test drag performance, now we don't need it anymore.

**Review by kiryldz — APPROVED** [review]
LGTM

File: extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt [hunk 10]
```diff
@@ -344,27 +357,42 @@ class GeoJsonSource(builder: Builder) : Source(builder.sourceId) {
...
-            setProperty(property, throwRuntimeException = false)
-            listener.invoke(this)
+      // No parseJob is working now, parse the data directly.
+      if (parsingJob == null) {
+        parseData(data, listener)
+      } else {
+        parsingJob?.let {
+          if (it.isActive) {
+            // still have parse job working, save the current data to waitingData.
+            pendingData = Pair(data, listener)
```
Comment (alexander-kulikovskii) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:367]: we save only one data for future parsing? 
maybe we should save it in list (queue)

File: extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt [hunk 11]
```diff
@@ -41,22 +39,37 @@ class GeoJsonSource(builder: Builder) : Source(builder.sourceId) {
...
-    } ?: run { geoJsonParsed = true }
+    } ?: run {
+      geoJsonParsed = true
+    }
+  }
+
+  private fun parsePendingData() {
+    pendingData?.let {
+      // have data waiting to parse, continue parse this data
+      parseData(it.first, it.second)
```
Comment (alexander-kulikovskii) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:62]:
```suggestion
      parseData(data = it.first, listener = it.second)
```

File: extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt [hunk 12]
```diff
@@ -41,22 +39,37 @@ class GeoJsonSource(builder: Builder) : Source(builder.sourceId) {
...
-              it.onGeoJsonParsed(this)
-            }
+      parsingJob = scope.launch {
+        val property = getPropertyValue(it)
+        geoJsonParsed = true
+        // we set parsed data when sync setter was not called during background work
+        if (ignoreParsedGeoJsonRegistry[sourceId] == false) {
+          setProperty(property, throwRuntimeException = false)
+          onGeoJsonParsedListenerList.forEach {
+            it.onGeoJsonParsed(this@GeoJsonSource)
```
Comment (alexander-kulikovskii) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:49]: should we handle exceptions here? coroutine error handling works in a different way

File: extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt [hunk 13]
```diff
@@ -41,22 +39,37 @@ class GeoJsonSource(builder: Builder) : Source(builder.sourceId) {
...
-        val property = it.toPropertyValue()
-        mainHandler.post {
-          geoJsonParsed = true
-          // we set parsed data when sync setter was not called during background work
-          if (ignoreParsedGeoJsonRegistry[sourceId] == false) {
-            setProperty(property, throwRuntimeException = false)
-            onGeoJsonParsedListenerList.forEach {
-              it.onGeoJsonParsed(this)
-            }
+      parsingJob = scope.launch {
```
Comment (alexander-kulikovskii) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:42]: should we somehow stop (cancel) this job?

File: extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt (hunk 10 above)
Comment (Chaoba) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:367]: No need to parse the pending data if have new pending data coming, we only need to parse the last one.

File: extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt (hunk 13 above)
Comment (Chaoba) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:42]: There is no lifecycle for GeoJsonSouce could enable us to cancel the job at the right time.

File: extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt (hunk 12 above)
Comment (Chaoba) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/sources/generated/GeoJsonSource.kt:49]: We don't throw an exception but will show the error log in logcat.

**Review by kiryldz — CHANGES_REQUESTED** [review]
@Chaoba after some offline discussion with @alexander-kulikovskii some idea came to our mind. Could you please check with a profiler if we spawn an actual `GeoJsonSourceParseThread` that lives all the time (as it uses static dispatcher).

If it does then technically this PR does not differ from original one that spawns `STYLE_WORKER` handler thread but could bring in some coroutine corner cases.

**Comment by Chaoba** [comment]
After offline discursion, I will close this pr for now since we don't get too many pros by introducing coroutine and many have potential bugs.

**Pull request closed without merging**